contracts/test/umbra-hook-test.js

```diff
@@ -0,0 +1,407 @@
+const { ethers, web3, artifacts } = require('hardhat');
+const { expectEvent, expectRevert } = require('@openzeppelin/test-helpers');
+const { expect } = require('chai');
+const { argumentBytes } = require('./sample-data');
+const { sumTokenAmounts, signMetaWithdrawal } = require('./utils');
+
+const Umbra = artifacts.require('Umbra');
+const TestToken = artifacts.require('TestToken');
+const MockHook = artifacts.require('MockHook');
+const { toWei, fromWei, BN } = web3.utils;
+const AddressZero = ethers.constants.AddressZero;
+
+describe('Umbra Hooks', () => {
+  const ctx = {};
+  let owner, sender, receiver, acceptor, relayer;
+
+  // amount that will be configured as the eth toll
+  const toll = toWei('0.01', 'ether');
+
+  // token fee the relayer will charge for meta-tx withdrawal
+  const relayerTokenFee = toWei('2', 'ether');
+
+  // The ethers wallet that will be used when testing meta tx withdrawals
+  const metaWallet = ethers.Wallet.createRandom();
+
+  // helper function to mint test tokens
+  const mintToken = async (toAddr, amount) => {
+    await ctx.token.mint(toAddr, toWei(amount));
+  };
+
+  // helper function to send test tokens to stealth address
+  const sendToken = async (fromAddr, stealthAddr, amount) => {
+    const toll = await ctx.umbra.toll();
+    const weiAmount = toWei(amount);
+
+    await ctx.token.approve(ctx.umbra.address, weiAmount, { from: fromAddr });
+
+    await ctx.umbra.sendToken(stealthAddr, ctx.token.address, weiAmount, ...argumentBytes, {
+      from: fromAddr,
+      value: toll,
+    });
+  };
+
+  // helper function to mint and send tokens to steatlh address
+  const mintAndSendToken = async (fromAddr, stealthAddr, amount) => {
+    await mintToken(fromAddr, amount);
+    await sendToken(fromAddr, stealthAddr, amount);
+  };
+
+  before(async () => {
+    [owner, sender, receiver, acceptor, relayer] = await web3.eth.getAccounts();
+    ctx.chainId = await web3.eth.getChainId();
+  });
+
+  beforeEach(async () => {
```

Comment:
Another nit: Do you mind using additional `describe` blocks to group the below tests where it makes sense to? A bit hard to follow the different combinations of hook/no hook, data/no data, meta-tx/regular-tx, etc. I think grouping might make it easier to make sure all the combinations are covered

Reply:
Good idea!